File: test/DubinsTest.cpp
#include "../src/maths.hh"
#include "../src/utils.hh"
#include <cmath>

#define FORMULA

Tuple<double> LSL (Angle th0, Angle th1, double _kmax)
{
#if 1
  double C=th0.cos()-th1.cos();
  double S=2*_kmax-th0.sin()+th1.sin();
  
  double temp1=2+4*pow2(_kmax)-(th0-th1).cos()-4*_kmax*(th0.sin()-th1.sin());
  
  if (temp1<0){
    return Tuple<double> (0);
  }
  
  double invK=1/_kmax;
  Angle sc_s1=Angle(th0.get()-atan2(C,S), Angle::RAD)*invK;
  double sc_s2=invK*sqrt(temp1);
  Angle sc_s3=Angle(atan2(C,S)-th1.get(), Angle::RAD)*invK;
  
  return Tuple<double> (3, sc_s1.get(), sc_s2, sc_s3.get());
  
#else
  
  double invK = 1/_kmax;
  double C = th1.cos()-th0.cos();
  double S = 2*_kmax + th0.sin()-th1.sin();
  
  Angle temp1 (atan2(C,S), Angle::RAD);
  double sc_s1 = invK*(temp1-th0).get();
  
  double temp2 = 2+4*pow2(_kmax) -2*(th0-th1).cos()+4*_kmax*(th0.sin()-th1.sin());
  if (temp2<0){
    return Tuple<double>(0);
  }
  double sc_s2 = invK * sqrt(temp2);
  double sc_s3 = invK*(th1-temp1).get();
  
  return Tuple<double> (3, sc_s1, sc_s2, sc_s3);
#endif
}

Tuple<double> RSR (Angle th0, Angle th1, double _kmax)
{
#if 1
  double C=th0.cos()-th1.cos();
  double S=2*_kmax-th0.sin()+th1.sin();
  
  double temp1=2+4*pow2(_kmax)-(th0-th1).cos()-4*_kmax*(th0.sin()-th1.sin());
  
  if (temp1<0){
    return Tuple<double> (0);
  }
  
  double invK=1/_kmax;
  Angle sc_s1=Angle(th0.get()-atan2(C,S), Angle::RAD)*invK;
  double sc_s2=invK*sqrt(temp1);
  Angle sc_s3=Angle(atan2(C,S)-th1.get(), Angle::RAD)*invK;
  
  return Tuple<double> (3, sc_s1.get(), sc_s2, sc_s3.get());
#else
  
  double invK = 1/_kmax;
  double C = th0.cos()-th1.cos();
  double S = 2*_kmax - th0.sin()+th1.sin();
  
  Angle temp1 (atan2(C,S), Angle::RAD);
  double sc_s1 = invK*(th0-temp1).get();
  
  double temp2 = 2+4*pow2(_kmax) -2*(th0-th1).cos()-4*_kmax*(th0.sin()-th1.sin());
  if (temp2<0){
    return Tuple<double>(0);
  }
  double sc_s2 = invK * sqrt(temp2);
  double sc_s3 = invK*(temp1-th1).get();
  
  return Tuple<double> (3, sc_s1, sc_s2, sc_s3);
#endif
}

Tuple<double> LSR (Angle th0, Angle th1, double _kmax)
{
#if 1
  double C = th0.cos()+th1.cos();
  double S=2*_kmax+th0.sin()+th1.sin();
  
  double temp1=-2+4*pow2(_kmax)+2*(th0-th1).cos()+4*_kmax*(th0.sin()+th1.sin());
  if (temp1<0){
    return Tuple<double> (0);
  }
  
  double invK=1/_kmax;
  
  double sc_s2=invK*sqrt(temp1);
  Angle sc_s1= Angle(atan2(-C,S)-atan2(-2, _kmax*sc_s2)-th0.get(), Angle::RAD)*invK;
  Angle sc_s3= Angle(atan2(-C,S)-atan2(2, _kmax*sc_s2)-th1.get(), Angle::RAD)*invK;
  
  return Tuple<double>(3, sc_s1.get(), sc_s2, sc_s3.get());
#else
  double invK = 1/_kmax;
  double C = th0.cos()+th1.cos();
  double S = 2*_kmax + th0.sin()+th1.sin();
  
  double temp1=atan2(-C,S);
  
  double temp2 = 4*pow2(_kmax) - 2 + 2*(th0-th1).cos() + 4*_kmax * (th0.sin() + th1.sin());
  if (temp2<0){
    return Tuple<double>(0);
  }
  double sc_s2 = invK * sqrt(temp2);
  double temp3 = -atan2(-2, sc_s2*_kmax);
  double sc_s1 = invK * Angle((temp3-th0.get()+temp1), Angle::RAD).get();
  double sc_s3 = invK * Angle((temp3-th1.get()+temp1), Angle::RAD).get();
  
  return Tuple<double> (3, sc_s1, sc_s2, sc_s3);
#endif
}

Tuple<double> RSL (Angle th0, Angle th1, double _kmax)
{
#if 1
  double C = th0.cos()+th1.cos();
  double S=2*_kmax-th0.sin()-th1.sin();
  
  double temp1=-2+4*pow2(_kmax)+2*(th0-th1).cos()-4*_kmax*(th0.sin()+th1.sin());
  if (temp1<0){
    return Tuple<double> (0);
  }
  
  double invK=1/_kmax;
  
  double sc_s2=invK*sqrt(temp1);
  Angle sc_s1= Angle(th0.get()-atan2(C,S)+atan2(2, _kmax*sc_s2), Angle::RAD)*invK;
  Angle sc_s3= Angle(th1.get()-atan2(C,S)+atan2(2, _kmax*sc_s2), Angle::RAD)*invK;
  
  return Tuple<double>(3, sc_s1.get(), sc_s2, sc_s3.get());
#else
  double invK = 1/_kmax;
  double C = th0.cos()+th1.cos();
  double S = 2*_kmax - th0.sin()-th1.sin();
  
  double temp1=atan2(C,S);
  
  double temp2 = 4*pow2(_kmax) - 2 + 2*(th0-th1).cos() + 4*_kmax*(th0.sin() + th1.sin());
  if (temp2<0){
    return Tuple<double>(0);
  }
  double sc_s2 = invK * sqrt(temp2);
  double temp3 = -atan2(-2, sc_s2*_kmax);
  double sc_s1 = invK * Angle((th0.get()-temp1+temp3), Angle::RAD).get();
  double sc_s3 = invK * Angle((th1.get()-temp1+temp3), Angle::RAD).get();
  
  return Tuple<double> (3, sc_s1, sc_s2, sc_s3);
#endif
}

Tuple<double> RLR (Angle th0, Angle th1, double _kmax)
{
#if 1
  double C=th0.cos()-th1.cos();
  double S=2*_kmax-th0.sin()+th1.sin();
  
  double temp1=0.125*(6-4*pow2(_kmax)+2*(th0-th1).cos()+4*_kmax*(th0.sin()-th1.sin()));
  
  if (std::abs(temp1)>1){
    return Tuple<double> (0);
  }
  
  double invK=1/_kmax;
  Angle sc_s2 = Angle(2*M_PI-acos(temp1), Angle::RAD)*invK;
  Angle sc_s1 = Angle(th0.get()-atan2(C, S)+0.5*_kmax*sc_s2.get(), Angle::RAD)*invK;
  Angle sc_s3 = Angle(th0.get()-th1.get()+_kmax*(sc_s2-sc_s1).get(), Angle::RAD)*invK;
  
  return Tuple<double>(3, sc_s1.get(), sc_s2.get(), sc_s3.get());
  
#else
  double invK = 1/_kmax;
  double C = th0.cos()-th1.cos();
  double S = 2*_kmax - th0.sin()+th1.sin();
  
  Angle temp1 (atan2(C,S), Angle::RAD);
  double temp2 = 0.125*(6 - 4*pow2(_kmax) + 2*(th0-th1).cos() + 4*_kmax*(th0.sin()-th1.sin()));
  if (std::abs(temp2)>1){
    return Tuple<double>(0);
  }
  
  double sc_s2 = invK*(Angle(2*M_PI-acos(temp2), Angle::RAD).get());
  double sc_s1 = invK*(th0-temp1+Angle(sc_s2*(0.5*_kmax), Angle::RAD)).get();
  double sc_s3 = invK*Angle(th1.get()-th0.get()+(sc_s2-sc_s1)*_kmax, Angle::RAD).get();
  
  return Tuple<double> (3, sc_s1, sc_s2, sc_s3);
#endif
}

Tuple<double> LRL (Angle th0, Angle th1, double _kmax)
{
#if 1
  double C=th1.cos()-th0.cos();
  double S=2*_kmax+th0.sin()-th1.sin();
  
  double temp1=0.125*(6-4*pow2(_kmax)+2*(th0-th1).cos()-4*_kmax*(th0.sin()-th1.sin()));
  
  if (std::abs(temp1)>1){
    return Tuple<double> (0);
  }
  
  double invK=1/_kmax;
  Angle sc_s2 = Angle(2*M_PI-acos(temp1), Angle::RAD)*invK;
  Angle sc_s1 = Angle(atan2(C, S)-th0.get()+0.5*_kmax*sc_s2.get(), Angle::RAD)*invK;
  Angle sc_s3 = Angle(th1.get()-th0.get()+_kmax*(sc_s2-sc_s1).get(), Angle::RAD)*invK;
  
  return Tuple<double>(3, sc_s1.get(), sc_s2.get(), sc_s3.get());
#else
  double invK = 1/_kmax;
  double C = th1.cos()-th0.cos();
  double S = 2*_kmax + th0.sin()-th1.sin();
  
  Angle temp1 (atan2(C,S), Angle::RAD);
  double temp2 = 0.125*(6 - 4*pow2(_kmax) + 2*(th0-th1).cos() - 4*_kmax*(th0.sin()-th1.sin()));
  
  if (std::abs(temp2)>1){
    return Tuple<double>(0);
  }
  double sc_s2 = invK*(Angle(2*M_PI-acos(temp2), Angle::RAD)).get();
  double sc_s1 = invK*(temp1-th0+Angle(sc_s2*(0.5*_kmax), Angle::RAD)).get();
  double sc_s3 = invK*Angle(th1.get()-th0.get()+(sc_s2-sc_s1)*_kmax, Angle::RAD).get();
  
  return Tuple<double> (3, sc_s1, sc_s2, sc_s3);
#endif
}

void printTuple (Tuple<double> T){
  printf("<");
  for (int i=0; i<T.size(); i++){
    if (i!=T.size()-1){
      printf("%f, ", T.get(i));
    }
    else {
      printf("%f", T.get(i));
    }
  }
  printf(">\n");
}

int main (){
  for (double th0=0.0; th0<=2*M_PI; th0+=0.1){
    for (double th1=0.0; th1<=2*M_PI; th1+=0.1){
      for (double kmax=0.0; kmax<=5; kmax+=0.1){
        Angle _th0=Angle(th0, Angle::RAD);
        Angle _th1=Angle(th1, Angle::RAD);
        printf("%f, %f, %f, ", th0, th1, kmax);
        printTuple(LSL(_th0, _th1, kmax));
        
        // printf("%f, %f, %f, ", th0, th1, kmax);
        // printTuple(LSR(_th0, _th1, kmax));
        
        // printf("%f, %f, %f, ", th0, th1, kmax);
        // printTuple(RSR(_th0, _th1, kmax));
        
        // printf("%f, %f, %f, ", th0, th1, kmax);
        // printTuple(RSL(_th0, _th1, kmax));
        
        // printf("%f, %f, %f, ", th0, th1, kmax);
        // printTuple(LRL(_th0, _th1, kmax));
        
        // printf("%f, %f, %f, ", th0, th1, kmax);
        // printTuple(RLR(_th0, _th1, kmax));
        
        // printf("\n");
      }
    }
  }
  return 0;
}

File: src/MatLab/dubinstest.m
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%% 
% Demo usage of Dubins shortest path function;
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

% Reset environment
close all; clear all; clc;

fl= fopen("/Users/enrico/GoogleDrive/Magistrale/1sem/Laboratory of Applied Robotics/LabRoboticsProject/data/test/DubinsTest_matlab.test", "w");

% value=0;
for th0=0.0 : 0.1 : 2*pi
  for th1=0.0 : 0.1 : 2*pi
    for kmax=0.0 : 0.1 : 5
      [ok, sc_s1, sc_s2, sc_s3]=LSL(th0, th1, kmax);
      if ok==true
        fprintf(fl, "%f, %f, %f, <%f, %f, %f>\n", th0, th1, kmax, sc_s1, sc_s2, sc_s3);
      else
        fprintf(fl, "<>\n");
      end

      [ok, sc_s1, sc_s2, sc_s3]=LSR(th0, th1, kmax);
      if ok==true
        fprintf(fl, "%f, %f, %f, <%f, %f, %f>\n", th0, th1, kmax, sc_s1, sc_s2, sc_s3);
      else
        fprintf(fl, "<>\n");
      end
      
      [ok, sc_s1, sc_s2, sc_s3]=RSR(th0, th1, kmax);
      if ok==true
        fprintf(fl, "%f, %f, %f, <%f, %f, %f>\n", th0, th1, kmax, sc_s1, sc_s2, sc_s3);
      else
        fprintf(fl, "<>\n");
      end
      
      [ok, sc_s1, sc_s2, sc_s3]=RSL(th0, th1, kmax);
      if ok==true
        fprintf(fl, "%f, %f, %f, <%f, %f, %f>\n", th0, th1, kmax, sc_s1, sc_s2, sc_s3);
      else
        fprintf(fl, "<>\n");
      end
      
      [ok, sc_s1, sc_s2, sc_s3]=LRL(th0, th1, kmax);
      if ok==true
        fprintf(fl, "%f, %f, %f, <%f, %f, %f>\n", th0, th1, kmax, sc_s1, sc_s2, sc_s3);
      else
        fprintf(fl, "<>\n");
      end

      [ok, sc_s1, sc_s2, sc_s3]=RLR(th0, th1, kmax);
      if ok==true
        fprintf(fl, "%f, %f, %f, <%f, %f, %f>\n", th0, th1, kmax, sc_s1, sc_s2, sc_s3);
      else
        fprintf(fl, "<>\n");
      end
      fprintf(fl, "\n");
    end
  end
end

% Normalize an angle (in range [0,2*pi))
function out = mod2pi(ang);
  out = ang;
  while (out < 0)
    out = out + 2 * pi;
  end
  while (out >= 2 * pi)
    out = out - 2 * pi;
  end
end

% LSL
function [ok, sc_s1, sc_s2, sc_s3] = LSL(sc_th0, sc_thf, sc_Kmax);
  invK = 1 / sc_Kmax;
  C = cos(sc_thf) - cos(sc_th0);
  S = 2 * sc_Kmax + sin(sc_th0) - sin(sc_thf);
  temp1 = atan2(C, S);
  sc_s1 = invK * mod2pi(temp1 - sc_th0);
  temp2 = 2 + 4 * sc_Kmax^2 - 2 * cos(sc_th0 - sc_thf) + 4 * sc_Kmax * (sin(sc_th0) - sin(sc_thf));
  if (temp2 < 0)
    ok = false; sc_s1 = 0; sc_s2 = 0; sc_s3 = 0;
    return;
  end
  sc_s2 = invK * sqrt(temp2);
  sc_s3 = invK * mod2pi(sc_thf - temp1);
  ok = true;
end

% RSR
function [ok, sc_s1, sc_s2, sc_s3] = RSR(sc_th0, sc_thf, sc_Kmax);
  invK = 1 / sc_Kmax;
  C = cos(sc_th0) - cos(sc_thf);
  S = 2 * sc_Kmax - sin(sc_th0) + sin(sc_thf);
  temp1 = atan2(C, S);
  sc_s1 = invK * mod2pi(sc_th0 - temp1);
  temp2 = 2 + 4 * sc_Kmax^2 - 2 * cos(sc_th0 - sc_thf) - 4 * sc_Kmax * (sin(sc_th0) - sin(sc_thf));
  if (temp2 < 0)
    ok = false; sc_s1 = 0; sc_s2 = 0; sc_s3 = 0;
    return;
  end
  sc_s2 = invK * sqrt(temp2);
  sc_s3 = invK * mod2pi(temp1 - sc_thf);
  ok = true;
end

% LSR
function [ok, sc_s1, sc_s2, sc_s3] = LSR(sc_th0, sc_thf, sc_Kmax);
  invK = 1 / sc_Kmax;
  C = cos(sc_th0) + cos(sc_thf);
  S = 2 * sc_Kmax + sin(sc_th0) + sin(sc_thf);
  temp1 = atan2(-C, S);
  temp3 = 4 * sc_Kmax^2 - 2 + 2 * cos(sc_th0 - sc_thf) + 4 * sc_Kmax * (sin(sc_th0) + sin(sc_thf));
  if (temp3 < 0)
    ok = false; sc_s1 = 0; sc_s2 = 0; sc_s3 = 0;
    return;
  end
  sc_s2 = invK * sqrt(temp3);
  temp2 = -atan2(-2, sc_s2 * sc_Kmax);
  sc_s1 = invK * mod2pi(temp1 + temp2 - sc_th0);
  sc_s3 = invK * mod2pi(temp1 + temp2 - sc_thf);
  ok = true;
end

% RSL
function [ok, sc_s1, sc_s2, sc_s3] = RSL(sc_th0, sc_thf, sc_Kmax);
  invK = 1 / sc_Kmax;
  C = cos(sc_th0) + cos(sc_thf);
  S = 2 * sc_Kmax - sin(sc_th0) - sin(sc_thf);
  temp1 = atan2(C, S);
  temp3 = 4 * sc_Kmax^2 - 2 + 2 * cos(sc_th0 - sc_thf) - 4 * sc_Kmax * (sin(sc_th0) + sin(sc_thf));
  if (temp3 < 0)
    ok = false; sc_s1 = 0; sc_s2 = 0; sc_s3 = 0;
    return;
  end
  sc_s2 = invK * sqrt(temp3);
  temp2 = atan2(2, sc_s2 * sc_Kmax);
  sc_s1 = invK * mod2pi(sc_th0 - temp1 + temp2);
  sc_s3 = invK * mod2pi(sc_thf - temp1 + temp2);
  ok = true;
end

% RLR
function [ok, sc_s1, sc_s2, sc_s3] = RLR(sc_th0, sc_thf, sc_Kmax);
  invK = 1 / sc_Kmax;
  C = cos(sc_th0) - cos(sc_thf);
  S = 2 * sc_Kmax - sin(sc_th0) + sin(sc_thf);
  temp1 = atan2(C, S);
  temp2 = 0.125 * (6 - 4 * sc_Kmax^2 + 2 * cos(sc_th0 - sc_thf) + 4 * sc_Kmax * (sin(sc_th0) - sin(sc_thf)));
  if (abs(temp2) > 1)
    ok = false; sc_s1 = 0; sc_s2 = 0; sc_s3 = 0;
    return;
  end
  sc_s2 = invK * mod2pi(2 * pi - acos(temp2));
  sc_s1 = invK * mod2pi(sc_th0 - temp1 + 0.5 * sc_s2 * sc_Kmax);
  sc_s3 = invK * mod2pi(sc_th0 - sc_thf + sc_Kmax * (sc_s2 - sc_s1));
  ok = true;
end

% LRL
function [ok, sc_s1, sc_s2, sc_s3] = LRL(sc_th0, sc_thf, sc_Kmax);
  invK = 1 / sc_Kmax;
  C = cos(sc_thf) - cos(sc_th0);
  S = 2 * sc_Kmax + sin(sc_th0) - sin(sc_thf);
  temp1 = atan2(C, S);
  temp2 = 0.125 * (6 - 4 * sc_Kmax^2 + 2 * cos(sc_th0 - sc_thf) - 4 * sc_Kmax * (sin(sc_th0) - sin(sc_thf)));
  if (abs(temp2) > 1)
    ok = false; sc_s1 = 0; sc_s2 = 0; sc_s3 = 0;
    return;
  end
  sc_s2 = invK * mod2pi(2 * pi - acos(temp2));
  sc_s1 = invK * mod2pi(temp1 - sc_th0 + 0.5 * sc_s2 * sc_Kmax);
  sc_s3 = invK * mod2pi(sc_thf - sc_th0 + sc_Kmax * (sc_s2 - sc_s1));
  ok = true;
end
